scopes/pipelines/builder/builder.graphql.ts

```diff
@@ -1,36 +1,123 @@
-import { Component } from '@teambit/component';
+import { Component, ComponentID } from '@teambit/component';
 import gql from 'graphql-tag';
-
+import { Scope } from '@teambit/legacy/dist/scope';
+import { ArtifactObject } from '@teambit/legacy/dist/consumer/component/sources/artifact-files';
+import isBinaryPath from 'is-binary-path';
 import { BuilderMain } from './builder.main.runtime';
+import { PipelineReport } from './build-pipeline-result-list';
+
+type ArtifactGQLFile = {
+  id: string;
+  name: string;
+  path: string;
+  content?: string;
+  downloadUrl?: string;
+};
+type ArtifactGQLData = ArtifactObject & { files: ArtifactGQLFile[]; componentId: ComponentID };
+type TaskReport = PipelineReport & {
+  artifact: ArtifactGQLData;
+  componentId: ComponentID;
+};
+const FILE_PATH_PARAM_DELIM = '~';
 
-// eslint-disable-next-line @typescript-eslint/no-unused-vars
-export function builderSchema(builder: BuilderMain) {
+export function builderSchema(builder: BuilderMain, scope: Scope) {
   return {
     typeDefs: gql`
-      type ExtensionDescriptor {
-        # extension ID.
-        id: String
+      type TaskReport {
+        id: String!
+        name: String!
+        description: String
+        startTime: String
+        endTime: String
+        errors: [String!]
+        warnings: [String!]
+        artifact(path: String): Artifact
+      }
 
-        # icon of the extension.
-        icon: String
+      type ArtifactFile {
+        # for GQL caching - same as the path
+        id: String!
+        name: String
+        path: String!
+        # artifact file content (only for text files). Use /api/<component-id>/~aspect/builder/<extension-id>/~<path> to fetch binary file data
+        content: String
+        downloadUrl: String
       }
 
-      type ExtensionArtifact {
-        # descriptor of the artifact's extension
-        extensionDescriptor: ExtensionDescriptor
+      type Artifact {
+        # artifact name
+        name: String
+        description: String
+        files: [ArtifactFile!]!
+      }
+
+      type AspectData {
+        # aspectId
+        id: String!
+        data: JSONObject
       }
 
       extend type Component {
-        # list of extension artifacts.
-        getArtifacts(hash: String!): [ExtensionArtifact]
+        pipelineReport(taskId: String): [TaskReport!]!
       }
     `,
+
     resolvers: {
       Component: {
-        // eslint-disable-next-line @typescript-eslint/no-unused-vars
-        getArtifacts: async (component: Component, { hash }: { hash: string }) => {
-          // const artifacts = await builder.getArtifacts(component.id, hash);
-          // return artifacts.map((artifact) => artifact.toObject());
+        pipelineReport: async (component: Component, { taskId }: { taskId?: string }) => {
+          const builderData = builder.getBuilderData(component);
```

Comment:
In general we should make this getBuilderData return data with ArtifactList from `builder/artifacts/artifact-list`
This is the new harmony type for artifacts.
Since this is a breaking change, we might want to create a new function that returns the new data, and deprecates the getBuilderData.
This true for many other APIs like
getArtifactsByName, getArtifactsByExtension, getArtifactsByExtensionAndName, getArtifactsByExtensionAndTaskName
(and maybe more)
if we can make the new APIs and use them at least here, it's a big step forward

Reply:
I started doing this here - https://github.com/teambit/bit/compare/master...storage-resolver in case you want to take a look
(It's old and contains many unrelated changes, but you can go directly to builder.main.runtime)